### PR TITLE
Fix/18554: Update reaction dynamically in the reaction popup

### PR DESCRIPTION
**Pull request by dukenv0307**
### Detail
We will move the logic to get information of the reaction to PopoverReactionList.js itself. So, this component could be dynamically updated

### Fixed Issues

$ https://github.com/Expensify/App/issues/18554
PROPOSAL: https://github.com/Expensify/App/issues/18554#issuecomment-1539941287

### Tests

1. On device 1 and login with User A and go to chat with user B
2. On device 2 and login with User B
3. On device 1 send a message to user B and react to that message. Long press on that emoji to see the emoji reaction number and the user who reacted it (Do not close the popup )
4. Now on device 2, from user B, react to the above message
5. Verify that the emoji reaction number and the users who reacted to the message are dynamically updated on the popup on device 1
- [x] Verify that no errors appear in the JS console

### Offline tests

None

### QA Steps

1. On device 1 and login with User A and go to chat with user B
2. On device 2 and login with User B
3. On device 1 send a message to user B and react to that message. Long press on that emoji to see the emoji reaction number and the user who reacted it (Do not close the popup )
4. Now on device 2, from user B, react to the above message
5. Verify that the emoji reaction number and the users who reacted to the message are dynamically updated on the popup on device 1
- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/129500732/bdb924c9-e752-4624-8610-09f375254c2b

</details>

<details>
<summary>Mobile Web - Safari && IOS</summary>

https://github.com/Expensify/App/assets/129500732/3608d553-8289-4e36-a07a-1a2a6ff763fa

</details>

<details>
<summary>Desktop</summary>

</details>

<details>
<summary>iOS</summary>

In the safari section

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/129500732/99045bdd-f017-4849-a8fa-a5d8d9dd8fc8

</details>

### CODE REVIEW

**Comment by dukenv0307** [comment]
@rushatgabhane everything is updated

**Comment by dukenv0307** [comment]
@rushatgabhane Sorry about my late. I just updated, help to check again

**Comment by dukenv0307** [comment]
@rushatgabhane Bump, please help to review

**Comment by strepanier03** [comment]
@rushatgabhane - Little bump for the PR review here the next time you're online and available. Thanks!

**Review by rushatgabhane — CHANGES_REQUESTED** [review]

**Review by rushatgabhane — DISMISSED** [review]
@cristipaval LGTM!

File: src/components/Reactions/ReportActionItemReactions.js [hunk 1]
```diff
@@ -49,6 +48,7 @@ const propTypes = {
      */
     // eslint-disable-next-line react/forbid-prop-types
     reactions: PropTypes.arrayOf(PropTypes.object).isRequired,
+    reportActionID: PropTypes.string.isRequired,
```
[the remark this answers is not in the document]
Comment (dukenv0307) [reply src/components/Reactions/ReportActionItemReactions.js:53]: Just updated comment like here https://github.com/Expensify/App/blob/896ad77e328e044df17f0df3b563f2b017369d82/src/pages/home/report/reportActionPropTypes.js#L4-L6

**Review by rushatgabhane — DISMISSED** [review]

**Review by rushatgabhane — APPROVED** [review]
LGTM! @cristipaval

**Comment by dukenv0307** [comment]
@cristipaval Friendly bump

**Review by cristipaval — APPROVED** [review]

File: src/pages/home/report/ReactionList/PopoverReactionList.js [hunk 2]
```diff
@@ -2,15 +2,58 @@ import React from 'react';
 import {Dimensions} from 'react-native';
 
 import lodashGet from 'lodash/get';
+import lodashMap from 'lodash/map';
+import lodashFilter from 'lodash/filter';
+import lodashFind from 'lodash/find';
+import lodashEach from 'lodash/each';
+import lodashIsEqual from 'lodash/isEqual';
```
Comment (tgolen) [inline src/pages/home/report/ReactionList/PopoverReactionList.js:9]: Using underscore is our standard to keep everything consistent: https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#object--array-methods

Please remove these and use underscore instead.

Comment (tgolen) [reply src/pages/home/report/ReactionList/PopoverReactionList.js:9]: `lodashGet` is the only one we allow since underscore doesn't have an equivalent.

File: src/pages/home/report/ReactionList/PopoverReactionList.js [hunk 3]
```diff
@@ -2,15 +2,58 @@ import React from 'react';
...
+
+/**
+ * Given an emoji object and a list of senders it will return an
+ * array of emoji codes, that represents all used variations of the
+ * emoji.
+ * @param {{ name: string, code: string, types: string[] }} emoji
+ * @param {Array} users
+ * @return {string[]}
+ * */
+const getUniqueEmojiCodes = (emoji, users) => {
```
Comment (tgolen) [inline src/pages/home/report/ReactionList/PopoverReactionList.js:45]: This method is now defined twice and needs to be DRYed up https://github.com/Expensify/App/blob/main/src/components/Reactions/ReportActionItemReactions.js#L24

File: src/pages/home/report/ReactionList/PopoverReactionList.js [hunk 4]
```diff
@@ -198,5 +317,15 @@ class PopoverReactionList extends React.Component {
...
 PopoverReactionList.propTypes = propTypes;
+PopoverReactionList.defaultProps = defaultProps;
 
-export default withLocalize(PopoverReactionList);
+export default compose(
+    withLocalize,
+    withOnyx({
+        reportActions: {
+            key: ({reportID}) => `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${reportID}`,
+            canEvict: false,
```
Comment (tgolen) [inline src/pages/home/report/ReactionList/PopoverReactionList.js:327]: Why is this here?

Comment (dukenv0307) [reply src/pages/home/report/ReactionList/PopoverReactionList.js:327]: @rushatgabhane  The PR is merged, so do we create a new PR to fix it or revert this PR?

Comment (dukenv0307) [reply src/pages/home/report/ReactionList/PopoverReactionList.js:327]: This is an explanation about canEvict field
https://github.com/Expensify/react-native-onyx/blob/1d496b001f327c509131597faf260a81c002be77/lib/withOnyx.js#L113-L116
@tgolen @rushatgabhane  I see that `canEvict: false` is added when we get reportAction everywhere in App repo. So I think in here we also add it for consistency

File: src/pages/home/report/ReactionList/PopoverReactionList.js (hunk 2 above)
Comment (dukenv0307) [reply src/pages/home/report/ReactionList/PopoverReactionList.js:9]: @tgolen @rushatgabhane I just created PR to fix: https://github.com/Expensify/App/pull/19183

File: src/pages/home/report/ReactionList/PopoverReactionList.js (hunk 4 above)
Comment (tgolen) [reply src/pages/home/report/ReactionList/PopoverReactionList.js:327]: I'm pretty sure the eviction logic is completely broken, and I think this should just be removed, but maybe it should be there for report actions... I do see it used everywhere. @vitHoracek @marcaaron do you have any guidance on the `canEvict` setting for report actions?

Comment (marcaaron) [reply src/pages/home/report/ReactionList/PopoverReactionList.js:327]: Ok, I did some poking and all the places where it is being used look kosher to me. There is documented guidance here: https://github.com/expensify/app#storage-eviction

Open to improvements or suggestions. I think there's _probably_ a better version of storage eviction waiting to be built at some point in all of our imaginations - but this one has gonna us pretty far. Let me know if there's anything else I can help with 🙇‍♂️

Comment (tgolen) [reply src/pages/home/report/ReactionList/PopoverReactionList.js:327]: @marcaaron The reason I believe that the eviction logic is broken is from this line here: https://github.com/Expensify/react-native-onyx/blob/main/lib/withOnyx.js#L129

When I was working with Thaibault (during ExpensiconX) and looking at migrating Onxy to TypeScript, we found that `mapping.connectionID` is `undefined`. This is because:

- The `connectionID` gets saved to the mapping [here](https://github.com/Expensify/react-native-onyx/blob/main/lib/Onyx.js#L676) in `Onyx.connect()`
- We assume that the `mapping` there is a reference to the same mapping object that is in `withOnyx` but actually, it's a copy as we can see by the spread operator being used on `mapping` [here](https://github.com/Expensify/react-native-onyx/blob/main/lib/withOnyx.js#L151) when it is passed to `Onyx.connect()`
- `checkEvictableKeys()` then looks through all the mappings in `mapOnyxToState` [here](https://github.com/Expensify/react-native-onyx/blob/main/lib/withOnyx.js#LL116C24-L116C38) and references `mapping.connectionID`. This is where it breaks 💥 because that object never had `connectionID` saved to it, the `connectionID` was only saved to a copy of it.

File: src/pages/home/report/ReactionList/PopoverReactionList.js [hunk 5]
```diff
@@ -81,23 +159,63 @@ class PopoverReactionList extends React.Component {
...
+                emojiName: '',
+                emojiCount: 0,
+            };
+        }
+        const emojiCount = selectedReaction.users.length;
+        const reactionUsers = lodashMap(selectedReaction.users, (sender) => sender.accountID.toString());
+        const emoji = lodashFind(emojis, (e) => e.name === selectedReaction.emoji);
+        const emojiCodes = getUniqueEmojiCodes(emoji, selectedReaction.users);
+        const hasUserReacted = Report.hasAccountIDReacted(this.props.currentUserPersonalDetails.accountID, reactionUsers);
+        const users = PersonalDetailsUtils.getPersonalDetailsByIDs(reactionUsers);
```
Comment (abdulrahuman5196) [inline src/pages/home/report/ReactionList/PopoverReactionList.js:197]: We missed to pass the accountID as the second param during this refractor causing this issue - https://github.com/Expensify/App/issues/20267
